certora/specs/AccessControlDefaultAdminRules.spec

```diff
@@ -0,0 +1,484 @@
+import "methods/IAccessControl.spec"
+import "AccessControl.spec"
+
+use rule onlyGrantCanGrant filtered {
+  f -> f.selector != acceptDefaultAdminTransfer().selector
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Helpers                                                                                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+function max_uint48() returns mathint {
+    return (1 << 48) - 1;
+}
+
+function nonZeroAccount(address account) returns bool {
+  return account != 0;
+}
+
+function timeSanity(env e) returns bool {
+  return
+    e.block.timestamp > 0 && // Avoids 0 schedules
+    e.block.timestamp + defaultAdminDelay(e) < max_uint48();
+}
+
+function intervalSanity(env e1, env e2) returns bool {
+  return e1.block.timestamp < e2.block.timestamp;
+}
+
+function delayChangeWaitSanity(env e, uint48 newDelay) returns bool {
+  return e.block.timestamp + delayChangeWait_(e, newDelay) < max_uint48();
+}
+
+function isSet(uint48 schedule) returns bool {
+  return schedule != 0;
+}
+
+function hasPassed(env e, uint48 schedule) returns bool {
+  return schedule < e.block.timestamp;
+}
+
+function min(uint48 a, uint48 b) returns mathint {
+  return a < b ? a : b;
+}
+
+function increasingDelaySchedule(env e, uint48 newDelay) returns mathint {
+  return e.block.timestamp + min(newDelay, defaultAdminDelayIncreaseWait());
+}
+
+function decreasingDelaySchedule(env e, uint48 newDelay) returns mathint {
+  return e.block.timestamp + defaultAdminDelay(e) - newDelay;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: defaultAdmin holds the DEFAULT_ADMIN_ROLE                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant defaultAdminConsistency(address account)
+  defaultAdmin() == account <=> hasRole(DEFAULT_ADMIN_ROLE(), account) 
+  {
+    preserved {
+      // defaultAdmin() returns the zero address when there's no default admin
+      require nonZeroAccount(account);
+    }
+  }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: Only one account holds the DEFAULT_ADMIN_ROLE                                                            │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant singleDefaultAdmin(address account, address another)
+  hasRole(DEFAULT_ADMIN_ROLE(), account) && hasRole(DEFAULT_ADMIN_ROLE(), another) => another == account
+  filtered { f -> f.selector != acceptDefaultAdminTransfer().selector }
```

Comment:
I don't understand. We are talking about the fact there is suppose to be only one account that is the default admin. Its essential we verify that `acceptDefaultAdminTransfer` doesn't break that invariant.

What your invariant is currently saying is :
> we know that if accounts are both admin, then they are the same, because there is only a single admin ...
> ... but we only check that this is true until someone calls `acceptDefaultAdminTransfer`.

So implicitly you are saying 
> if someone calls `acceptDefaultAdminTransfer`, then anything can happen, there might be multiple admins

Its ok to say "when processing the function, I'm going to help you with some `preserve`", but IMO its not ok to just say "lets not consider this function"

Reply:
I've spent the whole day on this and I don't find an easy way to make this invariant applicable to the `acceptDefaultAdminTransfer()`, I've tried with different ghost and hooks setups, as long as different variations of the same invariant, an example of how complex (and still not passing) is:

```
ghost defaultAdmins() returns uint256;

hook Sstore _roles[KEY bytes32 role].(offset 0)[KEY address account] bool granted (bool oldGranted) STORAGE {
  mathint countChange = 
    granted && !oldGranted ? 1 : // Granted
    !granted && oldGranted ? -1 : // Revoked
    0; // Anything else
    
  havoc defaultAdmins assuming defaultAdmins@new() == defaultAdmins@old() + countChange;
}

...

invariant singleDefaultAdmin(address account, address another)
  defaultAdmins() <= 1 &&
  (hasRole(DEFAULT_ADMIN_ROLE(), account) && hasRole(DEFAULT_ADMIN_ROLE(), another) => another == account)
  {
    preserved {
      requireInvariant singleDefaultAdmin(another, account);
      requireInvariant defaultAdminConsistency(defaultAdmin());
      require nonZeroAccount(account);
    }
  }
```

The main reason for it not pass is that I don't find a way of forcing the contract not to consider cases in which `role[DEFAULT_ADMIN_ROLE].members` only have one member at most since the beginning. The prover assumes this is a valid initial state:

<img width="573" alt="Captura de pantalla 2023-05-03 a la(s) 19 03 29" src="https://user-images.githubusercontent.com/33379285/235988339-0f2a04e1-2dee-4233-8406-0ab14d36119f.png">